### PR TITLE
Decorators for Polymer 3.0

### LINKED ISSUE #13: Test Safari and Edge
Our integration test only covers Chrome and Firefox, we should test Safari and Edge too.

### PR DESCRIPTION
**Pull request by aomarks**
- Now works with Polymer 3.0. Dropped support for 2.0. Added 2.x maintenance branch.
- Will be distributed on NPM as `@polymer/decorators` instead of Bower.
- Uses all ES modules instead of `Polymer.decorators` global and HTML import.
- Dropped support for Metadata Reflection (see new FAQ item for why).
- Test Safari/IE/Edge with Sauce (fixes #13).

You probably want to review each commit separately.

cc #10 #21 @TimvdLippe @43081j

### CODE REVIEW

File: src/decorators.ts [hunk 1]
```diff
@@ -71,42 +69,16 @@ export interface PropertyOptions {
...
-interface Reflect {
-  hasMetadata?:
-      (metadataKey: string, proto: object, targetKey: string) => boolean;
-  getMetadata?: (metadataKey: string, proto: object, targetKey: string) =>
-      object | undefined;
-}
-
 function createProperty(
-    proto: ElementPrototype, name: string, options?: PropertyOptions): void {
+    proto: ElementPrototype, name: string, options?: Partial<PropertyOptions>):
```
Comment (43081j) [inline src/decorators.ts]: because we use `Partial` here, does that mean the required `type` further up becomes optional again?

File: README.md [hunk 2]
```diff
@@ -113,14 +94,9 @@ class MyElement extends Polymer.Element {
...
-options](https://www.polymer-project.org/2.0/docs/devguide/properties) object.
+options](https://www.polymer-project.org/3.0/docs/devguide/properties) object.
 All standard options are supported, except for `value`; use a property
-initializer instead.
-
-If the [Metadata Reflection API](#metadata-reflection-api) is configured, the
-`type` option (which determines how Polymer de-serializes Element attributes
-for this property) will be inferred from the TypeScript type and can be
-omitted.
+initializer instead. `type` is required.
```
Comment (TimvdLippe) [inline README.md]: Add the following note to make this a little bit more clear: 

> Note that the `type` should be a JavaScript constructor like the Polymer types, not the TypeScript type

File: src/decorators.ts (hunk 1 above)
Comment (aomarks) [reply src/decorators.ts]: No, `property` will still enforce the narrower type, it's `Partial` here because `computed` doesn't set `type`.

File: README.md (hunk 2 above)
Comment (aomarks) [reply README.md]: Done.

**Comment by aomarks** [comment]
> Changes in general SFTM. Given the amount of TS(lint) configuration, are you planning on integrating this into the monorepo? That should remove quite a lot of boilerplate that is already fixed in the monorepo.

Wasn't planning on it. It doesn't really feel like a "tool" to me, and there are no inter-dependencies, so doesn't seem like there's much to gain. Deduping TS config doesn't really seem worth it to me.

File: README.md [hunk 3]
```diff
@@ -113,14 +94,10 @@ class MyElement extends Polymer.Element {
...
+options](https://www.polymer-project.org/3.0/docs/devguide/properties) object.
 All standard options are supported, except for `value`; use a property
-initializer instead.
-
-If the [Metadata Reflection API](#metadata-reflection-api) is configured, the
-`type` option (which determines how Polymer de-serializes Element attributes
-for this property) will be inferred from the TypeScript type and can be
-omitted.
+initializer instead. `type` is required, and must be one of the Polymer property
+constructor types (`String`, `Object`, etc.).
```
Comment (justinfagnani) [inline README.md]: Polymer itself defaults the type to String, should we?

File: README.md [hunk 4]
```diff
@@ -266,109 +245,52 @@ class MyElement extends
...
   myProperty: string = 'foo';
 }
 
-customElements.define(MyElement.is, MyElement);
+customElements.define('my-element', MyElement);
 ```
 
 ### What are the performance costs?
-The additional JavaScript served for this library is aproximately 8 KB (4 KB
+The additional JavaScript served for this library is approximately 6 KB (2 KB
```
Comment (justinfagnani) [inline README.md]: I think it's pretty standard practice to only report the gzipped size now, but still say it's gzipped.

Is this minified too? If not, I'd run the .js file though `uglify -mc --toplevel`

File: src/declarative-event-listeners.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ * Google as part of the polymer project is also subject to an additional IP
+ * rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {dedupingMixin} from '@polymer/polymer/lib/utils/mixin.js';
+import {PolymerElement} from '@polymer/polymer/polymer-element.js';
+
+export interface SuperClassConstructor {
+  // tslint:disable-next-line:no-any Required for constructor signature.
+  new(...args: any[]): PolymerElement;
```
Comment (justinfagnani) [inline src/declarative-event-listeners.ts]: does TS 3.0 support `...args: unknown[]` for mixins?

Comment (aomarks) [reply src/declarative-event-listeners.ts]: Nope, it breaks. Also don't want to use `unknown` for now because of g3.

File: README.md (hunk 3 above)
Comment (aomarks) [reply README.md]: I feel like it's pretty easy to forget, especially for users who may previously have been relying on reflect metadata, who would now just silently get a different de-serialization behavior.

File: README.md (hunk 4 above)
Comment (aomarks) [reply README.md]: Now only reporting gzipped size. Tried to minify, but neither mishoo/UglifyJS2 nor babel/minify support ES modules (at least as far as I can tell after poking around at options for a while).

Comment (aomarks) [reply README.md]: Needed *uglify-es* for ES module support. Updated the README (0.6KB minified + gzipped).

**Comment by aomarks** [comment]
Updated `@polymer/polymer-decorators` -> `@polymer/decorators`  -- it's cleaner.
